quinn/src/connection.rs

```diff
@@ -863,39 +865,49 @@ where
     }
 
     fn drive_timer(&mut self, cx: &mut Context) -> bool {
-        let mut keep_going = false;
-        loop {
-            if let Some(ref mut delay) = self.timer {
-                if delay.poll_unpin(cx) == Poll::Ready(()) {
-                    // We must get a fresh `now` each iteration. If the value were cached and tokio
-                    // deems a timer for a later time than the cached value to be expired, we'd get
-                    // stuck in an infinite loop resetting it.
-                    self.inner.handle_timeout(Instant::now());
-                    self.timer = None;
-                    // A timer expired, so the caller needs to check for new transmits, which might
-                    // cause new timers to be set.
-                    keep_going = true;
+        // Check whether we need to (re)set the timer. If so, we must poll again to ensure the
+        // timer is registered with the runtime (and check whether it's already
+        // expired).
+        match self.inner.poll_timeout().map(TokioInstant::from_std) {
+            Some(deadline) => {
+                if let Some(delay) = &mut self.timer {
+                    // There is no need to reset the tokio timer if the deadline
+                    // did not change
+                    if self
+                        .timer_deadline
+                        .map(|current_deadline| current_deadline != deadline)
+                        .unwrap_or(true)
+                    {
+                        delay.reset(deadline);
+                    }
+                } else {
+                    self.timer = Some(delay_until(deadline));
                 }
+                // Store the actual expiration time of the timer
+                self.timer_deadline = Some(deadline);
             }
-            // Check whether we need to (re)set the timer. If so, we must poll again to ensure the
-            // timer is registered with the runtime (and check whether it's already
-            // expired). Otherwise, exit the loop.
-            match (
-                self.inner.poll_timeout().map(TokioInstant::from_std),
-                &mut self.timer,
-            ) {
-                (Some(timeout), &mut None) => self.timer = Some(delay_until(timeout)),
-                (Some(timeout), &mut Some(ref mut delay)) if delay.deadline() != timeout => {
-                    delay.reset(timeout);
-                }
-                (None, _) => {
-                    self.timer = None;
-                    break;
-                }
-                _ => break,
+            None => {
+                self.timer_deadline = None;
+                return false;
             }
         }
-        keep_going
+
+        if let Some(_deadline) = self.timer_deadline {
```

Comment:
Let's invert the condition here to return early `if self.timer_deadline.is_none()`, and same for `delay.poll_unpin(cx).is_ready()`? That prevents the rightward drift here.

Reply:
Sounds good!

Reply:
I meant the second condition, too, not just the first.

Reply:
You mean

```
if delay.poll_unpin(cx).is_pending() {
    return false;
}
```

Can be done - but won't really save another nesting level here

Reply:
Yes, that's what I mean. It won't save another nesting level, but I think it will make the code easier to comprehend.